numba/cpython/setobj.py

```diff
@@ -1399,34 +1490,53 @@ def set_update(context, builder, sig, args):
 
     return context.get_dummy_value()
 
+def gen_operator_impl(op, impl):
+    @intrinsic
+    def _set_operator_intr(typingctx, a, b):
+        sig = a(a, b)
+        def codegen(context, builder, sig, args):
+            assert sig.return_type == sig.args[0]
+            impl(context, builder, sig, args)
+            return impl_ret_borrowed(context, builder, sig.args[0], args[0])
+        return sig, codegen
+
+    @overload(op)
+    def _ol_set_operator(a, b):
+        if all([isinstance(typ, types.Set) for typ in (a, b)]) and \
+                (a.dtype == b.dtype):
+            return lambda a, b: _set_operator_intr(a, b)
+
+
 for op_, op_impl in [
     (operator.iand, set_intersection_update),
     (operator.ior, set_update),
     (operator.isub, set_difference_update),
     (operator.ixor, set_symmetric_difference_update),
     ]:
-    @lower_builtin(op_, types.Set, types.Set)
-    def set_inplace(context, builder, sig, args, op_impl=op_impl):
-        assert sig.return_type == sig.args[0]
-        op_impl(context, builder, sig, args)
-        return impl_ret_borrowed(context, builder, sig.args[0], args[0])
+    gen_operator_impl(op_, op_impl)
 
 
 # Set operations creating a new set
 
-@lower_builtin(operator.sub, types.Set, types.Set)
-@lower_builtin("set.difference", types.Set, types.Set)
-def set_difference(context, builder, sig, args):
+@overload(operator.sub)
+@overload_method(types.Set, "difference")
+def impl_set_difference(a, b):
+    if not all([isinstance(typ, types.Set) for typ in (a, b)]):
```

Comment:
Think these also need `dtype` checks. Suggest checking all of these to make sure the `dtype` check is present.